### ISSUE TITLE
[Request] Support system dark mode

### ISSUE THREAD

**Issue by ThaumicMekanism** [mintty/wsltty#157]
This is something very minor but would it be possible to make the boarders of the terminal appear black if you have system dark mode turned on? I do not think this is already possible but would be pretty cool.

**Comment by mintty** [comment]
Mintty does not draw the "non-client area" (borders, title bar, scrollbar) itself. Windows does it.
If a Windows theme fails to apply its settings to windows, it's a Windows bug.
The best I find at the Windows API is SetSysColors but it has global effect.
See also mintty/mintty#751.

**Comment by mintty** [comment]
Marking invalid as this is a Windows deficiency rather than a mintty issue.

**Comment by kzeitz** [comment]
I see that you labeled this as "invalid" but I can't help but wonder how the stock Ubuntu terminal window seems to work properly when Windows uses dark-mode.

Maybe it's worth another look?

![image](https://user-images.githubusercontent.com/4561557/64912369-58623300-d6eb-11e9-85b9-938718904a4e.png)

**Comment by Biswa96** [comment]
@kzeitz It is magic of ConHost. Source code here https://github.com/microsoft/terminal

**Comment by mintty** [comment]
They have the internal know-how and the manpower to spend the extra effort to get this working.
I explained the invalid label above by the imposition that it's necessary at all to do extra work to get a system theme applied.

**Comment by mintty** [comment]
And you're welcome to take a look, and provide a patch.

**Comment by Biswa96** [comment]
Here is the patch [mintty_dark_mode.patch.txt](https://github.com/mintty/wsltty/files/3613138/mintty_dark_mode.patch.txt). I get it from microsoft/terminal repository. This uses one undocumented API from uxtheme.dll. I can make a PR.

**Comment by mintty** [comment]
Thanks for digging this out.
Isn't "DarkMode_Explorer" just one particular theme? Is this a generic solution? How can it be tested?
(I'm attaching a uniform-layout version of the patch, for use in a PR please.)
[mintty_dark_mode.patch2.txt](https://github.com/mintty/wsltty/files/3613157/mintty_dark_mode.patch2.txt)

**Comment by Biswa96** [comment]
Dont know the details. Just look the microsoft terminal repo and add the code in mintty. I shall find more details tomorrow 😴

**Comment by mintty** [comment]
When I configure "dark app mode", I can only see an effect on the scrollbar, not the title bar or other border (which are affected already without the patch).

**Comment by Biswa96** [comment]
Some queries:
* Should I use function pointer? Why not just use as an imported function?
* Which function type do you prefer, `-A` (ansi) or `-W` (unicode)?
*

**Comment by Biswa96** [comment]
Some queries:
* Should I use function pointer? Why not just use as an imported function?
* Which function type do you prefer, `-A` (ansi) or `-W` (unicode)?

**Comment by Biswa96** [comment]
For tittle bar, try to enable Settings (`Win + I`)  > Personalization > Colors > Tittle bar and window borders.

Some queries:
* Should I use function pointer for every GUI functions? Why not just directly linking that function to dll?
* Which function type do you prefer, `-A` (ansi) or `-W` (unicode)?

**Comment by mintty** [comment]
> For tittle bar, try to enable Settings (Win + I) > Personalization > Colors > Tittle bar and window 
borders.

This works already without the patch. Only difference with the patch is that the dark colour is now also applied to the scrollbar. (Which is good, of course.)

> Should I use function pointer for every GUI functions? Why not just directly linking that function to dll?

Mintty would not run in earlier Windows versions anymore.

> Which function type do you prefer, -A (ansi) or -W (unicode)?

Whenever there is a string argument, this should indeed be explicit. If the string is configurable and can contain non-ASCII, it should be W. If the string is a tag and certainly ASCII, it can be A.

> Warning: Found an issue. Right clicking on tittle bar does not show correct menu after applying the patch.

If you applied to the latest repository version, that's not due to your patch :)

**Comment by Biswa96** [comment]
Does dark tittle bar appear in cmd window without the setting change?

**Comment by Biswa96** [comment]
> This works already without the patch.

Does dark tittle bar appear in Command Prompt window without the setting change?

> Mintty would not run in earlier Windows versions anymore.

Then should I have to check if all the function pointer exist? Otherwise those will be null and mintty will crash.

> that's not due to your patch :)

Any hint?

**Comment by mintty** [comment]
>> This works already without the patch.

> Does dark tittle bar appear in Command Prompt window without the setting change?

Yes.

> >Mintty would not run in earlier Windows versions anymore.

> Then should I have to check if all the function pointer exist? Otherwise those will be null and mintty will crash.

Yes. Actually, SystemParametersInfoA is documented to be available since Windows 2000 (https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-systemparametersinfoa#requirements), so it is safe to use directly, without dynamic dll retrieval.
And I'm quite puzzled by GetProcAddress(uxtheme, MAKEINTRESOURCEA(132)); does that function really have no name?

> > that's not due to your patch :)

> Any hint?

https://github.com/mintty/mintty/commit/8a0e547c4a6aa9aba0f309f2a0102f3cdeafaef5

**Comment by Biswa96** [comment]
> does that function really have no name?

Welcome to undocumented Windows realm.

**Comment by Biswa96** [comment]
> I can only see an effect on the scrollbar, not the title bar or other border

I found the solution with some reverse engineering. The code is highly undocumented. Replace `19` value with `20`.  The dark mode can also be set with `SetWindowCompositionAttribute()`.

**Comment by Biswa96** [comment]
> I can only see an effect on the scrollbar, not the title bar or other border

I found the solution with some reverse engineering. The code is highly undocumented. Replace `19` value with `20`.  The dark mode can also be set with `SetWindowCompositionAttribute()`. Also check if the colors change when System settings is changed to light mode and vice versa.

Here is my proposed patch file: 
[mintty_whole_dark_mode.patch.txt](https://github.com/mintty/wsltty/files/3624687/mintty_whole_dark_mode.patch.txt)

**Comment by mintty** [comment]
When I googled `struct ACCENTPOLICY`, I saw all 4 components to be int, not enum or long. That could be a problem, depending on compilation environment (e.g. 32/64 bit, cygwin/msys). Also `    struct WINCOMPATTRDATA` showed int rather than enum.

In your new patch, value 20 is not used.
I see no significant difference except that `DwmSetWindowAttribute` is *not* called, which makes no difference for me.

Right now, I rather feel inclined to remove the whole Glass feature, at least from the Options menu.

**Comment by Biswa96** [comment]
> I rather feel inclined to remove the whole Glass feature

Why? It is also a great feature. It's not related with dark mode. Should I make a PR?

**Comment by mintty** [comment]
Oh, I checked the wrong thing, sorry.
Yet I do not see any difference by your new patch. The current version (with your previous patch) makes the mintty scrollbar follow the theme, that's all and working fine.

**Comment by Biswa96** [comment]
In my case, mintty shows dark tittle bar and dark scrollbar both.

**Comment by Biswa96** [comment]
Perhaps, can anyone test that patch in their system?

**Comment by mintty** [comment]
> In my case, mintty shows dark tittle bar and dark scrollbar both.

Yes, if dark mode is selected and application to borders and title bar is activated in Windows.
Without the patch, it does not show the scrollbar dark. But it does show the borders and title bar already.

**Comment by Biswa96** [comment]
Isn't the issue is solved now?

**Comment by mintty** [comment]
Fixed as far as I see; I'll close after next release.

**Comment by mintty** [comment]
Released 3.0.3.

**Comment by matthew-nm** [comment]
After this change, is the titlebar supposed to be black like the default terminal? For me it is still white.
![Capture](https://user-images.githubusercontent.com/5170156/78259176-2f4f9700-74b1-11ea-998c-42c575038855.PNG)

**Comment by mintty** [comment]
In my tests, after configuring Windows as someone guided above (using dark theme mode or whatever), it made a difference and I got a dark title bar. I have no overview of exact dependencies.

**Comment by boikyeet** [comment]
I installed wsltty and the title bar was dark just as expected for dark mode. After updating windows to version 2004 the title bar is now white even though I'm still in dark mode

**Comment by mintty** [comment]
Did they introduce yet another tweak that applications need to apply just to get preferences applied?
If anyone has additional information, provide it here.

**Comment by adipose** [comment]
I created the dark theme for mpc-hc, and I use a windows api to enable the dark titlebar.  On windows 2004 it still works fine.  I also wrote a custom window that draws it from scratch.  However, upon finding the windows API I disabled that code (still in the source).

Below is the the critical code that makes it work.  This code is MFC but only a slight tweak would work for win32.  Note that WCA_USEDARKMODECOLORS = 26.

```cpp
        HMODULE hUser = GetModuleHandleA("user32.dll");
        if (hUser) {
            pfnSetWindowCompositionAttribute setWindowCompositionAttribute = (pfnSetWindowCompositionAttribute)GetProcAddress(hUser, "SetWindowCompositionAttribute");
            if (setWindowCompositionAttribute) {
                ACCENT_POLICY accent = { ACCENT_ENABLE_BLURBEHIND, 0, 0, 0 };
                WINDOWCOMPOSITIONATTRIBDATA data;
                data.Attrib = WCA_USEDARKMODECOLORS;
                data.pvData = &accent;
                data.cbData = sizeof(accent);
                setWindowCompositionAttribute(window->GetSafeHwnd(), &data);
            }
        }
```

**Comment by mintty** [comment]
Please explain the purpose of your comment. Code similar to yours is already active in mintty. Is there anything that does not work?

**Comment by adipose** [comment]
You mean other than it not working under 2004 and showing a white bar?

**Comment by mintty** [comment]
Yes thanks, you had said it would still work on 2004 (which I don't have yet), and something about your own software which did not appear to relate to mintty.
Anyway, if that modification is required to keep it working from 2004, I wonder how to handle this whole crap. It's a Windows bug that functionality is poorly described and then even stops working.

**Comment by mintty** [comment]
OK, I have Version 2004 but Build 19035, while https://en.wikipedia.org/wiki/Windows_10_version_history lists Build 19041 for 2004, so maybe I have a prerelease. Anyway, dark title bar still works for me.

**Comment by Berrysoft** [comment]
The dark title bar works with 19041, on my machine.
![Dark](https://user-images.githubusercontent.com/37586447/83114091-194e0380-a0fb-11ea-82fb-c3dd0c16fc00.png)

**Comment by mintty** [comment]
@adipose,
> I created the dark theme for mpc-hc

So maybe you created your own theme which looks dark but is not classified "dark" for Windows?
Please try the built-in Windows dark theme.

**Comment by adipose** [comment]
I just got 19042 and wsltty did not have a dark bar. I came here and found others stating the same. Yet, mpc-hc (unrelated project that I work on) continues to work. So I shared some info in case it could help.

Perhaps the difference is not in how the theme is implemented, but in how we detect it should be enabled. If so, rather than try to detect and match windows theme, maybe it should be a user selectable option.

Firefox and Chrome I think just enable it per user preferences rather than following Windows.

**Comment by adipose** [comment]
> @adipose,
> 
> > I created the dark theme for mpc-hc
> 
> So maybe you created your own theme which looks dark but is not classified "dark" for Windows?
> Please try the built-in Windows dark theme.
I use the windows api for the window theme, no custom code for that.

**Comment by mintty** [comment]
Does it work if you replace WCA_ACCENT_POLICY with 26 in struct WINCOMPATTRDATA data?

**Comment by adipose** [comment]
I only tried the latest release of wsltty. I didn't compile it. Maybe I need a newer build?

**Comment by adipose** [comment]
I have now compiled it, and using 26 does nothing by itself.  The "enabled" flag does not get set to true.  However, if I force it to be true, 26 looks better to me, as 19 leaves a blurry background of sorts.

**Comment by adipose** [comment]
OK, so after playing around I see what it's doing.  It's trying to use honor "glass" settings, and one of the things that does is make the titlebar dark when it's out of focus.  However, it doesn't do anything while it's focused, as many windows (cmd.exe, explorer.exe) do in Windows 10 and the dark theme is enabled.

I believe what is missing to support the dark theme properly in windows 10 is a check like this (this code I wrote for mpc-hc, fyi):

```cpp
        if (ERROR_SUCCESS == key.Open(HKEY_CURRENT_USER, _T("Software\\Microsoft\\Windows\\CurrentVersion\\Themes\\Personalize"), KEY_READ)) {
            DWORD useTheme = -1;
            if (ERROR_SUCCESS == key.QueryDWORDValue(_T("AppsUseLightTheme"), useTheme)) {
                if (0 == useTheme) {
                    bWindows10DarkThemeActive = true;
                }
            }
        }
```

If this is true then we need to enable the accent policy when the glass setting is off, with mode 26.

cmd.exe in foreground:

![image](https://user-images.githubusercontent.com/3324395/83294937-055ddc80-a1a3-11ea-8538-5316a06084bc.png)

mintty in foreground with glass disabled:

![image](https://user-images.githubusercontent.com/3324395/83295161-61286580-a1a3-11ea-8043-1785fd2dbcda.png)

If this is a different issue than this thread was about, I apologize.  I thought it was the same issue.

**Comment by mintty** [comment]
The "enabled" flag would be enabled by setting `Transparency=Glass`, an old deprecated setting with side effects that never appeared to work for me.
However, if the usage of SetWindowCompositionAttribute is decoupled and combined with various transparency values, behaviour becomes different and maybe useful. Glass mode suddenly works, possibly after some recent Windows update, I don't know.
However, my observations seem to differ slightly: I'm getting a blurry effect ("glass mode") only with ACCENT_ENABLE_BLURBEHIND and with WCA_ACCENT_POLICY (*not* WCA_USEDARKMODECOLORS); this is enforced even if `Transparency=none`.
The opposite setting, WCA_USEDARKMODECOLORS with ACCENT_DISABLED spoils dark mode in the title bar.
Combining WCA_USEDARKMODECOLORS and ACCENT_ENABLE_BLURBEHIND looks the same for me as none of them, except if the window is unfocussed, the background gets dimmed.

**Comment by adipose** [comment]
https://github.com/mintty/mintty/pull/1005

Here's a pull request that is working for me. Personally I think the glass mode is a worthless gimmick but some like it.  So in this patch, if glass is not enabled, I follow the Windows 10 dark mode setting.  Otherwise it should act as before when glass is on.

**Comment by adipose** [comment]
Also note that, as far as I know, nothing has changed since 1809 that would have broken this.  I've been using the same basic policy in mpc-hc the whole time and nothing has broken.

At times, when using mintty, the glass option didn't even show up for me.  Then, suddenly it surfaced again.  I'm not sure what happened. Maybe you must restart after some changes to see that option.

**Comment by adipose** [comment]
@mintty , what is your setting for theme?  I have seen some people try to "fake" dark mode by using accent colors and then choosing a color very close to black.  This puts a "blackish" titlebar on most apps.  You can't choose absolute black because for some reason that's not allowed.

This is what I call Windows 10 Dark mode/theme:

![image](https://user-images.githubusercontent.com/3324395/83305200-23ccd380-a1b5-11ea-8c5a-cb1a7cf2f4b8.png)

**Comment by adipose** [comment]
Finally, to make the dark mode title bar, DwmExtendFrameIntoClientArea is not required, but I think it is for glass.  So that should probably be disabled, too, when glass is disabled.

**Comment by Berrysoft** [comment]
> I believe what is missing to support the dark theme properly in windows 10 is a check like this

That has been done by Uxtheme.dll, which is used to determine dark theme by mintty. I added that code.

**Comment by Berrysoft** [comment]
@adipose So... have you solved this problem?

**Comment by adipose** [comment]
Latest release of wsltty did not work. After rebuilding from source, it works. Sorry for the misinformation.

**Comment by Berrysoft** [comment]
Then it seems that wsltty doesn't include the latest version of mintty. Maybe it's a new problem, I think.

**Comment by mintty** [comment]
The two recent tweaks are in fact not yet in the wsltty release version:
  * Tweak title bar dark mode adaptation to undocumented Windows changes (#983, mintty/wsltty#157).
  * Support dark mode for menus (#983, mintty/wsltty#157).

So maybe they solve the issue.

**Comment by mintty** [comment]
Released 3.2.0.
